cdc/sinkv2/eventsink/cloudstorage/dml_worker.go

```diff
@@ -0,0 +1,304 @@
+					if err != nil {
+						d.errCh <- err
+						return
+					}
+
+					for _, cb := range callbacks {
+						if cb != nil {
+							cb()
+						}
+					}
+					log.Debug("write file to storage success", zap.Int("workerID", d.id),
+						zap.String("namespace", d.changeFeedID.Namespace),
+						zap.String("changefeed", d.changeFeedID.ID),
+						zap.String("schema", tbl.Schema),
+						zap.String("table", tbl.Table),
+						zap.String("path", path),
+					)
+				}
+			}
+		}
+	}()
+}
+
+// backgroundDefragmentMsgs defragments eventFragments for each active table.
+func (d *dmlWorker) backgroundDefragmentMsgs(ctx context.Context) {
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case table := <-d.activeTablesCh.Out():
+				if atomic.LoadUint64(&d.isClosed) == 1 {
+					return
+				}
+				v, ok := d.defragmenters.Load(table)
+				if !ok {
+					log.Panic("failed to load defragmenter for table",
+						zap.Int("workerID", d.id),
+						zap.String("namespace", d.changeFeedID.ID),
+						zap.String("changefeed", d.changeFeedID.ID),
+						zap.String("schema", table.Schema),
+						zap.String("table", table.Table),
+					)
+				}
+				defrag := v.(*defragmenter)
+				v, ok = d.msgChannels.Load(table)
+				if !ok {
+					log.Panic("failed to load msg channel for table",
+						zap.Int("workerID", d.id),
+						zap.String("namespace", d.changeFeedID.ID),
+						zap.String("changefeed", d.changeFeedID.ID),
+						zap.String("schema", table.Schema),
+						zap.String("table", table.Table),
+					)
+				}
+				msgChan := v.(*chann.Chann[*common.Message])
+				written, err := defrag.writeMsgs(ctx, msgChan)
+				if err != nil {
+					log.Error("dml worker failed to defragment messages",
+						zap.Int("workerID", d.id),
+						zap.String("namespace", d.changeFeedID.ID),
+						zap.String("changefeed", d.changeFeedID.ID),
+						zap.String("schema", table.Schema),
+						zap.String("table", table.Table),
+						zap.Error(err))
+
+					d.errCh <- err
+					return
+				}
+				log.Debug("dml worker defragments messages success",
+					zap.Int("workerID", d.id),
+					zap.String("namespace", d.changeFeedID.ID),
+					zap.String("changefeed", d.changeFeedID.ID),
+					zap.String("schema", table.Schema),
+					zap.String("table", table.Table),
+					zap.Uint64("written", written),
+				)
+			}
+		}
+	}()
+}
+
+// backgroundDispatchMsgs dispatch eventFragment to each table's defragmenter,
+// and it periodically emits flush tasks.
+func (d *dmlWorker) backgroundDispatchMsgs(ctx context.Context, ch *chann.Chann[eventFragment]) {
+	var readyTables []versionedTable
+	ticker := time.NewTicker(d.flushInterval)
+
+	d.wg.Add(1)
+	go func() {
+		log.Debug("dml worker started", zap.Int("workerID", d.id),
+			zap.String("namespace", d.changeFeedID.Namespace),
+			zap.String("changefeed", d.changeFeedID.ID))
+		defer d.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
```

Comment:
Only the ticker can trigger `flush`es? Or we also can `flush` based on pending message count?

Reply:
And if tasks are scheduled based on tables, the ticker can be based on tables, too. I think it's helpful to reduce S3 RPC calls.

Reply:
In the latest commit, the table file size excess can also trigger the flush task.